homeassistant/helpers/script.py

```diff
@@ -223,288 +666,59 @@ def referenced_entities(self):
-            self.async_run(variables, context), self.hass.loop
+            self.async_run(variables, context), self._hass.loop
         ).result()
 
     async def async_run(
         self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
-
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
-            try:
-                await self._handle_action(action, variables, context)
-            except _SuspendScript:
-                # Store next step to take and notify change listeners
-                self._cur = cur + 1
-                if self._change_listener:
-                    self.hass.async_add_job(self._change_listener)
-                return
-            except _StopScript:
-                break
-            except Exception:
-                # Store the step that had an exception
-                self._exception_step = cur
-                # Set script to not running
-                self._cur = -1
-                self.last_action = None
-                # Pass exception on.
-                raise
-
-        # Set script to not-running.
-        self._cur = -1
-        self.last_action = None
-        if self._change_listener:
-            self.hass.async_add_job(self._change_listener)
+    ) -> bool:
+        """Run script."""
+        if self.is_running:
+            if self._if_running == IF_RUNNING_IGNORE:
+                self._log("Skipping script")
+                return False
+
+            if self._if_running == IF_RUNNING_ERROR:
+                self._raise("Already running")
+            if self._if_running == IF_RUNNING_RESTART:
+                self._log("Restarting script")
+                self.async_stop()
 
-    def stop(self) -> None:
-        """Stop running script."""
-        run_callback_threadsafe(self.hass.loop, self.async_stop).result()
+        self.last_triggered = utcnow()
+        if self._run_mode == RUN_MODE_LEGACY:
+            if self.is_running:
+                shared = cast(Optional[_LegacyScriptRun], self._runs[0])
+            else:
+                shared = None
+            run: _ScriptRunBase = _LegacyScriptRun(
+                self._hass, self, variables, context, shared
+            )
+        else:
+            if self._run_mode == RUN_MODE_BACKGROUND:
+                run = _BackgroundScriptRun(self._hass, self, variables, context)
+            else:
+                run = _BlockingScriptRun(self._hass, self, variables, context)
+        self._runs.append(run)
+        await run.async_run()
+        return True
```

Comment:
why this ?

Reply:
Sorry, not sure what you're referring to. Do you mean why the new return value?

If so, this is preparatory to the changes coming to the `automation` integration. Given the new `if_running` options, it will need to know if the call to `Script.async_run` actually started a new "run" so it can determine if it should update its `last_triggered` attribute. You'll notice that just above, if `self._if_running` is `IF_RUNNING_IGNORE` it returns `False`.

Reply:
Maybe we can put this in when we add it to the automation integration?

Not to get ahead on future PRs, but I think that a script deciding not to run should not impact if an automation triggered or not. It triggered, it's action just got ignored.

Reply:
I can take it out for now.

But don't confuse the `script` integration and the script helper in this context. 😉 This return value (from the script helper) is telling the caller (i.e., `automation`) if the `action` "sequence" ran, not if that sequence happened to have a call to a `script` in it (that decided not to run.)

`homeassistant/helpers/script.py` should probably be called `homeassistant/helpers/sequence.py`, but that's probably too big of a change.

BTW, the automation's `last_triggered` attribute is already a misnomer, IMHO, and has caused a lot of confusion (as I've seen on the forum.) E.g., it is not updated if a `trigger` "fires" but a `condition` prevents the `action` sequence from running. It is only updated when the `action` sequence actually runs. It should probably be called `last_action_run`, but that would probably be too big of a change, too.

I was planning to keep the behavior of that attribute the same.